stdlib/mexpr/decision-points.mc

```diff
@@ -4,34 +4,56 @@ include "digraph.mc"
 include "string.mc"
 include "ast-builder.mc"
 include "eq-paths.mc"
+include "prelude.mc"
 
 -- This file contains implementations related to decision points. In particular,
 -- it implements:
 -- * A language fragment for decision points (HoleAst)
 -- * An algorithm for AST -> call graph conversion (Ast2CallGraph fragment)
 -- * Program transformations for programs with decision points (HolyCallGraph)
 
+let _top = nameSym "top"
+
+let _eqn = lam n1. lam n2.
+  if and (nameHasSym n1) (nameHasSym n2) then
+    nameEqSym n1 n2
+  else
+    error "Name without symbol."
+
 lang HoleAst
   syn Expr =
   | TmHole {tp : Type,
-            startGuess : Dyn,
+            startGuess : Expr,
             depth : Int}
+  sem symbolize (env : Env) =
 end
 
 lang HoleAstPrettyPrint = HoleAst + TypePrettyPrint
-  sem pprintCode (indent : Int) =
-  | TmHole h -> strJoin "" ["Hole (", getTypeStringCode indent h.tp, ", ",
-                                      pprintCode indent h.startGuess, ", ",
-                                      pprintCode indent h.depth, ")"]
+  sem isAtomic =
+  | TmHole _ -> false
+
+  sem pprintCode (indent : Int) (env : Env) =
+  | TmHole h ->
+    match pprintCode indent env h.startGuess with (env1, startStr) then
+      match pprintCode indent env1 h.depth with (env2, depthStr) then
+        (env2,
+          join ["Hole (",
+                strJoin ", " [getTypeStringCode indent h.tp, startStr, depthStr],
+                ")"])
+      else never
+    else never
 end
 
 -- Temporary workaround: uniquely labeled decision points
```

Comment:
Is this due to some transformation needed pre-symbolize and thus not accurate any more, or is it referring to something else and still relevant?

Reply:
I need the hole and application AST nodes to have unique symbols, symbolize gives unique symbols to identifiers only (or did I misunderstand something?).

Reply:
No, I think that's correct

Reply:
The test suite takes 14s for me when running with this PR, compared to around 4s before. @lingmar can you take a closer look at the test cases that you have added in this PR? I would prefer if the testing time does not increase significantly. Thanks!

Reply:
@david-broman I run basically the same tests on the same 17 ASTs as before. I did some detective work and currently  each AST takes ~0.6 s to test, while previously the time was ~0.2 s per AST. The majority of the runtime is spent in the `transform` function, i.e. it seems like `transform` just got ~3x slower! I am really surprised by this since the transformation is almost exactly like before. Part of the slowdown lies in that I have to run `symbolize` on the transformed AST (disabling this drops the runtime to ~0.4 s per AST). Another part, I suspect, comes from running `symbolize` on the test-ASTs (before the transform), however I couldn't easily test this hypothesis as I need them to be symbolized in order to run the test cases.

To decrease the runtime, perhaps I can do the transformation in a symbolize-preserving way so that I don't have to run `symbolize` on the transformed AST (tricky but possible), and/or remove some of the ASTs from the test cases.